src/execution/index/art/art.cpp

```diff
@@ -204,6 +204,148 @@ void ART::GenerateKeys(DataChunk &input, vector<unique_ptr<Key>> &keys) {
 	}
 }
 
+struct BuildARTOffsets {
+	BuildARTOffsets(idx_t start_p, idx_t end_p, data_t byte_p) : start(start_p), end(end_p), byte(byte_p) {};
+	idx_t start;
+	idx_t end;
+	data_t byte;
+};
+
+void ART::Build(vector<unique_ptr<Key>> &keys, row_t *row_ids, Node *&node, idx_t start_idx, idx_t end_idx,
+                idx_t depth) {
+
+	// test if we reached a leaf
+	auto prefix_start = depth;
+	while (keys[start_idx]->len != depth && keys[start_idx]->data[depth] == keys[end_idx]->data[depth]) {
+		depth++;
+	}
+
+	// we reached a leaf
+	if (depth == keys[start_idx]->len) {
+
+		// end_idx is inclusive
+		auto num_row_ids = end_idx - start_idx + 1;
+
+		// check for possible constraint violation
+		if ((IsUnique() || IsPrimary()) && num_row_ids != 1) {
+			throw ConstraintException("New data contains duplicates on indexed column(s)");
+		}
+
+		// new row ids of this leaf
+		auto new_row_ids = unique_ptr<row_t[]>(new row_t[num_row_ids]);
+		for (idx_t i = 0; i < num_row_ids; i++) {
+			new_row_ids[i] = row_ids[start_idx + i];
+		}
+
+		node = new Leaf(*keys[start_idx], prefix_start, move(new_row_ids), num_row_ids);
+
+	} else { // create a new node and recurse
+
+		// we will find at least two offset entries, otherwise we'd have reached a leaf
+		vector<BuildARTOffsets> offsets;
+		idx_t offset = start_idx;
+		for (idx_t i = start_idx + 1; i <= end_idx; i++) {
+			if (keys[i - 1]->data[depth] != keys[i]->data[depth]) {
+				BuildARTOffsets entry(offset, i - 1, keys[i - 1]->data[depth]);
+				offsets.push_back(entry);
+				offset = i;
+			}
+		}
+		BuildARTOffsets entry(offset, end_idx, keys[end_idx]->data[depth]);
+		offsets.push_back(entry);
+
+		if (offsets.size() < 5) {
+			node = new Node4();
```

Comment:
Bit of a nitpick here, but can you change this to
```c++
		if (offsets.size() <= 4) {
 			node = new Node4();
```

And for the 16/48 as well? Makes it more readable (at least for me) :)

Reply:
I think this should be changed to a `NodeType Node::GetTypeBySize(uint8_t size)` that returns the NodeType and then care use the CreateNode function.